app/models/user.rb

```diff
@@ -2,9 +2,70 @@ class User < ApplicationRecord
   has_many :posts, dependent: :destroy
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships, foreign_key: "creator_id", dependent: :destroy
+  has_many :inverse_friendships, class_name: "Friendship", foreign_key: "recipient_id", dependent: :destroy
   belongs_to :location, required: false
+  validates :email, uniqueness: true, presence: true
+  validates :name, presence: true
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: %i[facebook]
+
+
+  def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
+      end
+    end
+  end
+
+  def confirmed_friends
+    created_friendships = Array.new
+    created_friendships = friendships.select { |i| i.status == true}.map do |f|
+      f.recipient
+    end
```

Comment:
- [ ] `**[OPTIONAL]**` we can use `confirmed_friendships` relation that we have created earlier to find the friends with [`has_many :through`](https://guides.rubyonrails.org/association_basics.html#the-has-many-through-association). You can see more examples on how to use it here [1](https://stackoverflow.com/questions/11600928/when-should-one-use-a-has-many-through-relation-in-rails/11601221#11601221).
So we can have another relation: `has_many: friends, through: confirmed_friendships`. Now you can just say `user.friends` and it will return a db relation with all the friends. Also it will become much easier to confirm a friendship with `friends << friend`.
The main difference from your approach is that your `confirmed_friends` return an array which is not the best way to deal with data in rails application [ActiveRecord Relations are not Arrays](http://succinctsoftware.com/2013/04/06/activerecord-relations-are-not-arrays/). Also see [this](https://stackoverflow.com/questions/17331862/converting-an-array-of-objects-to-activerecordrelation) on other developers struggling to convert array to database relations as their application grows.